Refactoring TestRunner class

krmahadevan opened the pull request:
* Extracted out un-related methods to their 
own utility classes.
* Removed deprecated TestRunner constructors
* Added unit tests.

Fixes # .

### Did you remember to?

- [ ] Add test case(s)
- [ ] Update `CHANGES.txt`

We encourage pull requests that:

* Add new features to TestNG (or)
* Fix bugs in TestNG

If your pull request involves fixing SonarQube issues then we would suggest that you please discuss this with the 
[TestNG-dev](https://groups.google.com/forum/#!forum/testng-dev) before you spend time working on it.

krmahadevan commented on src/main/java/org/testng/GuiceHelper.java:
```diff
@@ -0,0 +1,69 @@
+    }
+
+
+    private Module[] getModules(Guice guice, Injector parentInjector, Class<?> testClass) {
```
fixed.

krmahadevan commented on src/main/java/org/testng/GuiceHelper.java:
```diff
@@ -0,0 +1,69 @@
+import java.util.List;
+
+class GuiceHelper {
+    private ITestContext context;
```
fixed.

krmahadevan commented on src/main/java/org/testng/TestRunner.java:
```diff
@@ -353,46 +272,27 @@ private void initListeners() {
-      if (listener == null) {
-        listener = ClassHelper.newInstance(c);
-      }
+      ITestNGListener listener = TestListenerHelper.createListener(factory, c);
```
fixed.

krmahadevan commented on src/main/java/org/testng/TestRunner.java:
```diff
@@ -570,59 +458,8 @@ private void addMetaGroup(String name, List<String> groupNames) {
-    }
-
-    return result;
+    return GroupsHelper.createGroups(groups, m_metaGroups);
```
I didn't understand this comment. Can you please elaborate ?

juherr replied:
Just a style issue: `createGroups` takes local var then class attribute, and `createListenerFactory` takes class attribute then local var.

krmahadevan replied:
ok, changed.

krmahadevan commented on src/main/java/org/testng/TestRunner.java:
```diff
@@ -1039,201 +723,14 @@ private void afterRun() {
         "Running test " + m_testName + " on " + m_classMap.size() + " " + " classes, "
-        + " included groups:[" + mapToString(m_xmlMethodSelector.getIncludedGroups())
-        + "] excluded groups:[" + mapToString(m_xmlMethodSelector.getExcludedGroups()) + "]");
+        + " included groups:[" + Strings.mapToString(m_xmlMethodSelector.getIncludedGroups())
```
fixed.

krmahadevan commented on src/main/java/org/testng/TestRunner.java:
```diff
@@ -1523,50 +1001,18 @@ public void addListener(ITestNGListener listener) {
-  private final Collection<InvokedMethod> m_invokedMethods = new ConcurrentLinkedQueue<>();
+  private final Collection<IInvokedMethod> m_invokedMethods = new ConcurrentLinkedQueue<>();
 
   private void dumpInvokedMethods() {
```
We are using it from within `org.testng.TestRunner#afterRun` to dump this information on the console, whenever the user sets the verbosity as 4 or higher.

juherr replied:
The verbosity level logic should be there IMO

krmahadevan replied:
fixed.

krmahadevan commented on src/main/java/org/testng/internal/AbstractParallelWorker.java:
```diff
@@ -0,0 +1,92 @@
+        private ClassMethodMap classMethodMap;
+        private List<IClassListener> listeners;
+        private ITestContext testContext;
+        private IAnnotationFinder finder;
```
I did consider that, but it kinda would look weird, having a class with so many parameters as args for the constructor. I am trying to refrain from introducing methods/constructors with that many number of parameters. Let me see if I can introduce a builder instead.

krmahadevan replied:
fixed this by introducing a builder.

krmahadevan commented on src/main/java/org/testng/internal/ClassBasedParallelParallelWorker.java:
```diff
@@ -0,0 +1,136 @@
+        List<IWorker<ITestNGMethod>> result = Lists.newArrayList();
+        // Methods that belong to classes with a sequential=true or parallel=classes
+        // attribute must all be run in the same worker
+        Set<Class> sequentialClasses = Sets.newHashSet();
```
fixed.

krmahadevan commented on src/main/java/org/testng/internal/ClassBasedParallelParallelWorker.java:
```diff
@@ -0,0 +1,136 @@
+            if (sequentialClasses.contains(c)) {
+                if (!processedClasses.contains(c)) {
+                    processedClasses.add(c);
+                    if (System.getProperty("experimental") != null) {
```
I haven't used it. So I would like to first use it, and see what it has to offer. If it's useful, maybe we can enrich the documentation etc., for it so that people can use it. If its not useful, I will remove it as part of a new PR. I have raised an issue which will capture this https://github.com/cbeust/testng/issues/1605

juherr replied:
👍

krmahadevan commented on src/main/java/org/testng/internal/ClassBasedParallelParallelWorker.java:
```diff
@@ -0,0 +1,136 @@
+                attributes.getListeners());
+    }
+
+    private List<MethodInstance> methodsToMultipleMethodInstances(ITestNGMethod... sl) {
```
fixed.

krmahadevan commented on src/main/java/org/testng/internal/ClassBasedParallelParallelWorker.java:
```diff
@@ -0,0 +1,136 @@
+        return vResult;
+    }
+
+    private Collection<List<IMethodInstance>> createInstances(List<IMethodInstance> methodInstances) {
```
fixed.

krmahadevan commented on src/main/java/org/testng/internal/DynamicGraphHelper.java:
```diff
@@ -0,0 +1,209 @@
+import java.util.List;
+import java.util.Map;
+
+public class DynamicGraphHelper {
```
fixed.

krmahadevan commented on src/main/java/org/testng/internal/DynamicGraphHelper.java:
```diff
@@ -0,0 +1,209 @@
+
+    private static ListMultiMap<ITestNGMethod, ITestNGMethod> createClassDependencies(
+            ITestNGMethod[] methods, XmlTest test) {
+        Map<String, List<ITestNGMethod>> classes = Maps.newHashMap();
```
It is being used in the next couple of lines in the for loop iteration.

krmahadevan commented on src/main/java/org/testng/internal/InstanceBasedParallelParallelWorker.java:
```diff
@@ -0,0 +1,37 @@
+class InstanceBasedParallelParallelWorker extends AbstractParallelWorker {
+    @Override
+    public List<IWorker<ITestNGMethod>> createWorkers(Arguments arguments) {
+        List<IWorker<ITestNGMethod>> result = Lists.newArrayList();
```
fixed.

krmahadevan commented on src/main/java/org/testng/internal/MethodHelper.java:
```diff
@@ -335,6 +339,64 @@ private static String calculateMethodCanonicalName(Method m) {
     return result;
   }
 
+  public static void fixMethodsWithClass(ITestNGMethod[] methods,
```
Cannot do this, because this method is mostly being invoked from values extracted from `org.testng.ITestClass` wherein the methods are mostly being returned as an array.

krmahadevan commented on src/main/java/org/testng/internal/MethodHelper.java:
```diff
@@ -335,6 +339,64 @@ private static String calculateMethodCanonicalName(Method m) {
 
+  public static void fixMethodsWithClass(ITestNGMethod[] methods,
+                                   ITestClass testCls,
+                                   List<ITestNGMethod> methodList) {
```
This one is tricky, and will require more changes. I agree that it is not very intuitive to have a method update a parameter, but I dont want to complicate this PR by introducing a lot of complexity. I have added a `TODO` to fix this code smell.

krmahadevan commented on src/main/java/org/testng/internal/MethodHelper.java:
```diff
@@ -335,6 +339,64 @@ private static String calculateMethodCanonicalName(Method m) {
+  }
+
+  public static List<ITestNGMethod> invokedMethodsToMethods(Collection<IInvokedMethod> invokedMethods) {
+    List<ITestNGMethod> result= Lists.newArrayList();
```
fixed.

krmahadevan commented on src/test/java/org/testng/internal/DynamicGraphHelperTest.java:
```diff
@@ -0,0 +1,166 @@
+    public void testCreateDynamicGraphAllIndependent() {
+        DynamicGraph<ITestNGMethod> graph = newGraph(IndependentTestClassSample.class);
+        assertThat(graph.getFreeNodes()).hasSize(2);
+        for (Map.Entry<ITestNGMethod, List<DynamicGraph.Edge<ITestNGMethod>>> edge : graph.getEdges().entrySet()) {
```
fixed.

krmahadevan commented on src/main/java/org/testng/internal/ClassBasedParallelParallelWorker.java:
```diff
@@ -0,0 +1,136 @@
+        return result;
+    }
+
+    private static IMethodInstance[] findClasses(List<IMethodInstance> methodInstances, Class<?> c) {
```
fixed.

krmahadevan commented on src/main/java/org/testng/TestRunner.java:
```diff
@@ -1473,18 +961,8 @@ public void setVerbose(int n) {
-    if (listener instanceof ITestNGListener) {
-      addListener((ITestNGListener) listener);
-    }
+  void addTestListener(ITestListener listener) {
```
That change requires a bit more research. Introducing this change, breaks the build with a lot of failures. Now that this method is package only, I will get back to this. For now I am going to add a `TODO` marker.

juherr replied:
TODO is fine. I think we will have a similar thing on `TestNG`.

krmahadevan commented on src/main/java/org/testng/GuiceHelper.java:
```diff
@@ -0,0 +1,69 @@
+    }
+
+    public Injector getInjector(IClass iClass) {
+        Annotation annotation = AnnotationHelper.findAnnotationSuperClasses(Guice.class, iClass.getRealClass());
```
I didn't quite get this.  Can you please elaborate ?

juherr replied:
Maybe we can add something like:

```java
<A extends Annotation> A findAnnotationSuperClasses(Class<A> clazz, IClass iClass) {
  return AnnotationHelper.findAnnotationSuperClasses(clazz, iClass.getRealClass());
}
```

krmahadevan replied:
Do we really need to add that, especially when all we have to do is invoke `iClass.getRealClass()` to use the existing method that's already there ?

juherr replied:
You can keep it. It was just an idea :)

But if possible, can you fix the method signature?
From

```java
public static Annotation findAnnotationSuperClasses(Class<?> annotationClass, Class parameterClass)
```

to

```java
public static <A extends Annotation> A findAnnotationSuperClasses(Class<A> annotationClass, Class<?> parameterClass)
```

krmahadevan replied:
fixed, amended and changes pushed.

krmahadevan commented:
@juherr - I have fixed the review comments and pushed in the changes. Please have a look.

krmahadevan commented:
@juherr - Amended the commit and pushed the changes. Please review.

krmahadevan commented:
@juherr - I have addressed all observations. Please let me know if there's anything outstanding from my side to get this merged.

juherr commented:
@krmahadevan I will check again with all fixes asap

juherr commented:
@krmahadevan The build is failing on java7

krmahadevan commented:
@juherr - It was a compilation error specific to JDK7. Sigh! I have fixed it now. I am not too keen on installing JDK7 to find such issues. I wasn't able to detect this neither on the IDE (IntelliJ) nor via the command line when building using gradle (Am guessing that we would have set our JDK level to 7, but not sure why it still slips). Any idea how I can trap these sort of silly issues before I push code, but with using JDK7 itself ? 

Btw, I have fixed, amended and pushed the changes once again. The Travis builds pass as well. Please help take a look.

cbeust commented on src/main/java/org/testng/DefaultListenerFactory.java:
```diff
@@ -0,0 +1,10 @@
+package org.testng;
```
Maybe this should go in the `internal` package?

krmahadevan replied:
Fixed.

cbeust commented on src/main/java/org/testng/internal/AbstractParallelWorker.java:
```diff
@@ -0,0 +1,116 @@
+        if (XmlSuite.ParallelMode.INSTANCES.equals(mode)) {
+            return new InstanceBasedParallelParallelWorker();
+        }
+        return new ClassBasedParallelParallelWorker();
```
Two "parallel"?

krmahadevan replied:
Fixed. Silly me. Dont know how that one slipped through.

cbeust commented on src/main/java/org/testng/internal/AbstractParallelWorker.java:
```diff
@@ -0,0 +1,116 @@
+                instance = new Arguments();
+            }
+
+            public Builder usingMethods(List<ITestNGMethod> methods) {
```
A nitpick but I think `using` is overly verbose. For builders, I think it's fine to use straight property names:

    new Builder()
        .methods(...)
        .invoker(...)
        .build()

krmahadevan replied:
Fixed.

krmahadevan commented:
@cbeust - I fixed the review comments, rebased off of master and pushed in the changes. Please help take a look
